### PR TITLE
Tests: Integration test harness fixes

### PR DESCRIPTION
**Pull request by mic704b**
Some devs are reporting spurious errors when running integration tests.

https://discourse.joplinapp.org/t/integration-showallnotes-test-failing-randomly/6664
https://discourse.joplinapp.org/t/automated-testing-framework/6705

This PR contains modifications to resolve the issues.

## Random failures in integration tests
These issues are due to using timeouts to wait for asynchronous processing to complete in response a test action. On some devs PCs it seems these timeouts are not big enough, so the action is incomplete and the test fails. The modification is to wait for middleware processing to complete before proceeding with the tests. This is achieved by replacing the sleeps with with calls to `TestApp.waitForProcessing()`.

A second issue is due to notes having identical timestamps which can affect the sorting order.  To fix, a small sleep was inserted between creation consecutive test notes.

## "Accessing database before it has been initialised" errors
These are due to bugs in the tear down of the test app.  The fixes include
- redoing the shutdown of timers in `registry.js`. This is the most complex timer system and an error was made earlier.
- Adding `cancelTimer()` methods for `ResourceFetcher`, `ResourceService` and `RevisionService` . These were issed earlier.
- Refactor the passing of the database to use methods instead

Note that the changes to `registry.js` look far more extensive than they actually are. Due to wrapping the function in a `try` `finally` block, the code indentation changed, and git diff is not displaying it well.

## Revision unit test failures
Some tests in the Revision unit tests were failing, also suspected to be due to slow processing on some dev PC's.  The fix is to update the `services_Revision` unit tests to record and use exact timestamps rather than rely on timers.

The changes make the test execution deterministic and so more robust to differences in test machine processing speeds. I am assuming in making these changes, the purpose of the tests is to test app logic, and not to test minimum performance levels. ie we're trying to test the routines work, and not that they are fast.

For more discussion see [here](https://discourse.joplinapp.org/t/integration-showallnotes-test-failing-randomly/6664/47)

## Rename to "feature" tests
I have renamed the `integration` test files to be `feature` tests, to clearer communicate the scope of the test suites contained in the files.

### CODE REVIEW

**Comment by tessus** [comment]
The changes look great.

> the purpose of the tests is to test app logic, and not to test minimum performance levels.

I agree. Performance tests are a separate topic should that ever come up.

The only perf issues we have is the slow sync on Linux and the relatively long sync initialization. A sync without changes on macOS takes 5-9 seconds. 
But for narrowing that down we'd have to run network traces and use a proxy to retrieve the req/resp times.
The use of http/1.1 instead of h2 (on desktop) makes it also less performant than the mobile clients.

**Comment by mic704b** [comment]
Thanks tessus!

> Performance tests are a separate topic

I agree, and if we want to test them we should have their own dedicated test cases.

So in the meantime, when reviewing PRs we should look out for time dependencies in test cases.

File: ReactNativeClient/lib/services/ResourceFetcher.js [hunk 1]
```diff
@@ -206,25 +207,31 @@ class ResourceFetcher extends BaseService {
...
 
 	async autoAddResources(limit = null) {
-		if (limit === null) limit = 10;
+		this.autoAddResourcesCalls_.push(true);
+		try {
+			if (limit === null) limit = 10;
 
-		if (this.addingResources_) return;
-		this.addingResources_ = true;
+			if (this.addingResources_) return;
```
Comment (laurent22) [inline ReactNativeClient/lib/services/ResourceFetcher.js]: If we can return here, doesn't that mean that there will be a memory leak in your autoAddResourcesCalls_ array? (unless I missed something but the diff again is hard to follow)

File: CliClient/tests/test-utils.js [hunk 2]
```diff
@@ -439,6 +439,7 @@ async function createNTestFolders(n) {
 	for (let i = 0; i < n; i++) {
 		let folder = await Folder.save({ title: 'folder' });
 		folders.push(folder);
+		await time.msleep(10);
```
[the remark this answers is not in the document]
Comment (mic704b) [reply CliClient/tests/test-utils.js]: Just to ensure the created items have different timestamps.  Especially with notes, this can affect the sort order if the timestamps are the same, and so can cause random test failures.

Having identical timestamps is fine in the app, just not so good for testing.

File: ReactNativeClient/lib/BaseApplication.js [hunk 3]
```diff
@@ -332,6 +346,17 @@ class BaseApplication {
...
+	async waitForAutoAddResourcesFinished() {
+		return new Promise((resolve) => {
+			const iid = setInterval(() => {
+				if (!ResourceFetcher.instance().autoAddResourcesCalls_.length) {
+					clearInterval(iid);
+					resolve();
+				}
+			}, 100);
+		});
+	}
```
[the remark this answers is not in the document]
Comment (mic704b) [reply ReactNativeClient/lib/BaseApplication.js]: > Could that be part of ResourceFetcher?

Yes, it should be.  I was keeping it close to the `scheduleAutoAddResources()` which has the timer we are trying to stop.  I think they both probably belong in the `ResourceFetcher`.

> ResourceFetcher::waitForAllFinished() does?

Similar but not the same.  `waitForAllFinished()` is waiting on a broader set of conditions, where this method is specifically waiting just for the timer that is run in `BaseApplication`.

If we move `scheduleAutoAddResources()` to `ResourceFetcher` from `BaseApplication`, then I think it can be cleaned up better.  Would that be ok?

File: ReactNativeClient/lib/registry.js [hunk 4]
```diff
@@ -56,140 +56,152 @@ reg.syncTarget = (syncTargetId = null) => {
...
-	const synchronizer = await reg.syncTarget().synchronizer();
-	await synchronizer.waitForSyncToFinish();
-	await reg.scheduleSync(0);
+	reg.waitForReSyncCalls_.push(true);
+	try {
+		const synchronizer = await reg.syncTarget().synchronizer();
+		await synchronizer.waitForSyncToFinish();
+		await reg.scheduleSync(0);
+	} finally {
+		reg.waitForReSyncCalls_.pop();
```
[the remark this answers is not in the document]
Comment (mic704b) [reply ReactNativeClient/lib/registry.js]: If we are waiting to shutdown sync, and this method is also waiting (`on waitForSyncToFinish()`) then there could be a problem.

When sync finishes, if this method runs first before shutdown, then no problem.

But if it runs the other way round, then this would be a problem (we've pulled down the sync, but this method is using it).  So when shutting down, we give way so that this method can also complete before proceeding with the shutdown.

[the remark this answers is not in the document]
Comment (mic704b) [reply ReactNativeClient/lib/registry.js]: > unfortunately hard to follow the diff in this critical part of Joplin

Yes, sorry about that.  I needed to wrap the code in try/finally to avoid memory leaks, and the diff is a mess because of that.

File: ReactNativeClient/lib/services/ResourceFetcher.js (hunk 1 above)
Comment (mic704b) [reply ReactNativeClient/lib/services/ResourceFetcher.js]: No, I don't think so.  When we return within the `try` block, the `finally` will run, and so no memory leak.

**Comment by mic704b** [comment]
> I wasn't too sure about your xxxCalls_ pattern

It's actually your pattern!  It's orignally from `ItemChange`.

I was trying to implement the test harness changes so they had zero effect on the app itself, in that the shutdowns etc only every run in the test harness.

However, the only cross over is the `XXXCalls_` lists.  If there is ever a bug in them and they are unsymmetrical, then a `Calls_` list might grow indefinitely (memory leak).  This could affect the app.  For that reason, I was thinking of changing them to be a counter, instead of a push/pop list.

I've manually checked them and they look ok, but...

**Comment by mic704b** [comment]
> Perhaps later that could be used to shut down all services while the app is running

Yes, I didn't think of that!  Though at this point, I'm not confident it is completely watertight.

I use these changes on the Trash branch, where I have a lot more integration tests, and I haven't seen a problem.  But the relationships between the objects can be quite complex.  It might be a bit easier if we can dissolve some of the singletons along the way.

File: ReactNativeClient/lib/BaseApplication.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (mic704b) [reply ReactNativeClient/lib/BaseApplication.js]: Ok, I've moved it to `ResourceFetcher` and appears to fall into place nicely.

Be aware: this change involves a refactor of app code and not just code that runs during test.  I can revert if you're not comfortable with that.

**Comment by laurent22** [comment]
> It might be a bit easier if we can dissolve some of the singletons along the way.

What do you mean by this? To refactor them somehow into regular class/instances?

Otherwise I think this PR is good to merge. If you could just fix the conflict, feel free to merge afterwards.

**Comment by mic704b** [comment]
> What do you mean by this? To refactor them somehow into regular class/instances?

Yes. Singletons is good to allow for the app to self configure, but it makes difficult to tear down the  app in an orderly way (that we're trying to do here for integration testing). Just thinking out loud.
